### ISSUE TITLE
Add ttir.scatter op to the TTIR dialect

### ISSUE THREAD

**Issue by ajakovljevicTT** [tenstorrent/tt-mlir#1325]
As part of the effort to run stableHLO models on TT silicon, adding an implementation of the `scatter` op in the TTIR dialect.
This op currently has limited functionality, but enough to pass the current models, primarily due to the current limitations in tt-metal, for more information please see https://github.com/tenstorrent/tt-mlir/pull/1279.

**Comment by ajakovljevicTT** [comment]
Update: The op with limited functionality, as is currently implemented in metal (see issue https://github.com/tenstorrent/tt-metal/issues/4294) is now implemented in the TTIR dialect (https://github.com/tenstorrent/tt-mlir/pull/1279). 

However, as described above, we currently do not support custom indexes, nor custom function for merging the two tensor operands of the scatter, which is supported in StableHLO (https://github.com/tenstorrent/tt-mlir/issues/1278), so I'm leaving this issue open for now until we get full support.

**Comment by ajakovljevicTT** [comment]
@skhatriTT  The scatter with limited functionality was implemented, and this issue was left for when the there is more robust scatter support in ttnn. As far as I understand, this is currently being done, so maybe @sdjordjevicTT can give more details.

**Comment by svuckovicTT** [comment]
Looks like the time has come :)

TT-NN team have updated the scatter op, eliminating the old one. @brataTT implemented a [workaround](https://github.com/tenstorrent/tt-mlir/pull/4137/files#diff-dee09eb39a27c1dffb4e8e0e3a63b0f82a9859f6fa8f1e7de4786833a8cf99d3R13) in runtime, to support the old op.

I propose we rename the old scatter op to something else and implement the new one, in the same PR. That way, FEs can update their naming to the old one if they need it (tt-xla are the only ones using it right now), and have access to play around with the new one.

cc @tenstorrent/forge-developers-mlir-core

**Comment by sdjordjevicTT** [comment]
> I propose we rename the old scatter op to something else and implement the new one, in the same PR. That way, FEs can update their naming to the old one if they need it (tt-xla are the only ones using it right now), and have access to play around with the new one.


Yes, this plan sounds great. We should keep the old one and give some grace period to frontends to experiment with the changes.

**Comment by sdjordjevicTT** [comment]
Btw, @ajakovljevicTT, do you have some spare time to handle this, or should we continue from here?

**Comment by mrakitaTT** [comment]
@ajakovljevicTT is currently pretty occupied but @sdjukicTT will be able to take on this in a week or so, if you don't have anyone on your team that can focus on this right now @sdjordjevicTT. [This](https://github.com/tenstorrent/tt-mlir/issues/3567) scatter issue seems related and it blocks 6 of tt-xla models ([tt-xla issue](https://github.com/tenstorrent/tt-xla/issues/386)).

**Comment by ddilbazTT** [comment]
We are running into an unsupported scatter operation in `gpt-oss-20b` model:

```
%239 = "stablehlo.scatter"(%cst, %230, %238) <{scatter_dimension_numbers = #stablehlo.scatter<inserted_window_dims = [0, 1], scatter_dims_to_operand_dims = [0, 1], index_vector_dim = 2>}> ({
    ^bb0(%arg39: tensor<bf16>, %arg40: tensor<bf16>):
      stablehlo.return %arg40 : tensor<bf16>
    }) : (tensor<71x32xbf16>, tensor<71x4x2xi64>, tensor<71x4xbf16>) -> tensor<71x32xbf16>
```

or in ttir:
```
    %562 = "ttir.scatter"(%0, %540, %560, %561) <{index_vector_dim = 2 : i32, indices_are_sorted = false, input_batching_dims = array<i32>, inserted_window_dims = array<i32: 0, 1>, scatter_dims_to_operand_dims = array<i32: 0, 1>, scatter_indices_batching_dims = array<i32>, unique_indices = false, update_window_dims = array<i32>}> : (tensor<71x32xbf16>, tensor<71x4x2xi64>, tensor<71x4xbf16>, tensor<71x32xbf16>) -> tensor<71x32xbf16>
```

**Comment by sdjordjevicTT** [comment]
Will close this one as this one is not relevant anymore
